README
Win32-GUI-HyperLink
-------------------

VERSION

This is Win32::GUI::HyperLink v0.13

INTRODUCTION

Win32::GUI::HyperLink is a Win32::GUI class that implements a Label
object that acts as a clickable hyperlink:

By default it has a 'hand' Cursor, is drawn in blue text rather than
black and the text is underlined when the mouse is over the text.
The Label can be clicked to launch a hyperlink.

REQUIRES

Win32::GUI v0.99_1 or higher
Win32::API for full functionality - currently you lose the ability to
  have the text of the label dynamically underlined when the mouse is
  over the link if you don't have this module.
Win32::GUI::BitmapInline (as distributed with Win32::GUI).  This
  module is only needed if you don't have Win32::API.  If neither
  this module nor Win32:API are available, then the cursor will not
  change to a hand when the mouse is over the link

DOCUMENTATION

There is full POD documentation inside the HyperLink.pm file.
There is a demonstration script provided, showing most of the
features.

INSTALLATION

To install this module, run the following commands:

    perl Makefile.PL
    make
    make test
    make install

Alternatively, as a manual installation, extract the HyperLink.pm file
from the dirtribution, and put it at C:\Perl\site\lib\Win32\GUI
(adjust as necessary for your Perl installation directory)

A PPM distribution is available at http://www.robmay.me.uk/win32gui/

COPYRIGHT AND LICENCE

Copyright (C) 2005 Robert May

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

----------------------------------------------------------------------
$Id: README,v 1.4 2005/04/24 17:29:07 Robert May Exp $


Changes
Revision history for Win32-GUI-HyperLink

0.13     24 Aprl 2005
         Released on CPAN, and at http://www.robmay.me.uk/win32gui/
         as PPM

 - Corrected name of Hyperlink.pm to HyperLink.pm
 - Changed some private 'my' subs to regular subs to allow sub-classing
   to work properly.
 - Used Hook() to call event handlers, so that it works with both OEM
   and NEM.
 - Modified MouseIn and MouseOut event handlers to work with OEM.
 - Modified the way that API calls missing in Win32::GUI are checked,
   and only fallback to Win32::API when absolutely necessary
 - Updated the documentation.

0.12     28 February 2005
         Released on CPAN, and at http://www.robmay.me.uk/win32gui/
         as PPM

 - Removed left-over debug print statements
 - Moved under CVS - added keywords to all files; getting VERSION from
   release tag
 - Tidyied up demo script to correctly reference Win32::GUI::HyperLink
 - Removed Makefile.PL dependancy on Win32::GUI::BitmapInline, as it
   is always distributed with Win32::GUI
 - Removed Version from POD documentation, as I don't expect the
   documentation to change with version releases.

0.11     20 February 2005
         Released to perl-win32-gui-users for comment

 - Wrote initial test suite
 - Made API backward compatible with Win32::GUI::Hyperink v0.02
 - Removed dependency on Win32::API - now we use it for additional
   functionality if it is available, otherwise we manage without. 
   ShellExecute() is used if available (as it is faster and cleaner),
   but we fallback to using 'system start,$url'.  LoadCursor is used
   is available, but if not we fallback to an inlined cursor. 
   GetCapture is used if available - if it is not the the dynamic
   link underlining and onMouseIn/Out calls are not available.
 - Added fallback cursor using Win32::GUI::BitmapInline, in case
   we cannot load the default windows cursor for any reason (e.g.
   we don't have access to Win32::API).  If Win32::GUI::Bitmap::Inline
   is not available, then we simple don't change the cursor.
 - Correct VERSION format to x.xx for CPAN modules
 - removed Set/ReleaseCapture dependency on Win32::API, as these calls
   are already available in Win32::GUI 1.0
 - Added Win32::GUI::Window::AddHyperLink function to allow for
   Win32::GUI's standard $parent->AddControl interface.
 - Updated demo code to show new AddHyperLink usage
 - Turned comments into pod documentation, and completed documentation


0.10     15 February 2005
         First version by Robert May announced on perl-win32-gui-users
         mailing list for comment.

 - Complete re-write of code, original module only came to light after
   this release.

0.02    2001(?)
        Original version available on perl-win32-gui-users list

----------------------------------------------------------------------
$Id: Changes,v 1.3 2005/04/24 17:29:06 Robert May Exp $


Makefile.PL
# Makefile.PL for Win32::GUI::HyperLink
# $Id: Makefile.PL,v 1.4 2005/04/24 18:21:24 Robert May Exp $

use strict;
use warnings;
use ExtUtils::MakeMaker;

WriteMakefile(
    NAME                => 'Win32::GUI::HyperLink',
    AUTHOR              => 'Robert May <rmay@popeslane.clara.co.uk>',
    VERSION_FROM        => 'lib\Win32\GUI\HyperLink.pm',
    ABSTRACT_FROM       => 'lib\Win32\GUI\HyperLink.pm',
    PL_FILES            => {},
    PREREQ_PM => {
        'Test::More' => 0,
        'Win32::GUI' => 0,
        'Win32::API' => 0,
    },
    dist                => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean               => { FILES => 'Win32-GUI-HyperLink-*' },
);

# Add a new rule to make a ppm distribution
sub MY::postamble {
    return <<'MAKE_PPM';
ppm:
  $(RM_F) $(DISTVNAME)-PPM.zip
  make_ppm --arch=""
  $(ZIP) -9 $(DISTVNAME)-PPM.zip README-PPM.TXT $(DISTNAME).ppd $(DISTNAME).tar.gz
  $(RM_F) pod2htmld.tmp pod2htmli.tmp $(DISTNAME).ppd $(DISTNAME).tar.gz


MAKE_PPM
}


lib/Win32/GUI/demos/HyperLink/HyperLinkDemo.pl
# Demo on how to use Win32::GUI::HyperLink
#
# $Id: HyperLinkDemo.pl,v 1.3 2005/04/24 17:29:07 Robert May Exp $
#
# See the Win32::GUI::HyperLink POD documents for further
# Information
#
# Author: Robert May - rmay@popeslane.clara.co.uk
#
# Copyright (C) 2005 Robert May
#
# This script is free software; you can redistribute it and/or modify
# it under the same terms as Perl itself.
#

use strict;
use warnings;

use Win32::GUI;
#use Win32::GUI 1.0;

# The next line is for testing, and should be commented out in distributions
#use lib qw(../../../..);
use Win32::GUI::HyperLink;

# Some useful constants
sub CL_RED()    {0x0000FF}; # Red
sub CL_GREEN()  {0x00FF00}; # Green
sub UL_NEVER()  {0};        # Never underline link
sub UL_HOVER()  {1};        # underline link when the mouse is over the link (this is the default)
sub UL_ALWAYS() {2};        # Always underline link

my $title = "HyperLink Demo";

# A menu
my $Menu = Win32::GUI::MakeMenu(
  "&File"             => "File",
	"   > E&xit"        => { -onClick => sub { -1; } },
  "&Help"             => "Help",
	"   > &About ..."   => { -onClick => \&AboutWindow },
);

# A window with a menu
my $mw = Win32::GUI::Window->new(
	-title => $title,
	-menu => $Menu,
	-pos => [ 100, 100 ],
	-size => [ 240, 200 ],
  -resizable =>0,       # As I don't want to bother with re-drawing here!
  -maximizebox =>0,     # Ditto.
  -onTerminate => sub {-1},
);

# A status bar, where will might display the link information
my $status = $mw->AddStatusBar(
);

# Simplest usage
my $hl1 = $mw->AddHyperLink(
	-text => "http://www.robmay.me.uk/win32gui/",
	-pos => [10,10],
);

# Provide text to display in the label instead of the link,
# change the colour and add callbacks to put the link info
# into the status bar.  Never underlined.
my $hl2 = $mw->AddHyperLink(
  -text => "HyperLink.pm webpage",
  -url => "http://www.robmay.me.uk/win32gui/",
	-foreground => CL_RED,
	-pos => [10,30],
	-onMouseIn => \&setStatus,
	-onMouseOut => \&unsetStatus,
	-underline => UL_NEVER,
);

# A differnt font
my $font = Win32::GUI::Font->new(
	-name => "Comic Sans MS", 
	-size => 14,
);

# Using a different font: it really is just like a Label.
# Always underlined, and our own onClick handler
# Alternate constructor format
my $hl3 = Win32::GUI::HyperLink->new(
	$mw,
	-text => "email author",
	-url => 'mailto:rmay@popeslane.clara.co.uk',
	-foreground => CL_GREEN,
	-pos => [10,50],
	-font => $font,
	-onMouseIn => \&setStatus,
	-onMouseOut => \&unsetStatus,
	-onClick => \&hl3Click,
	-underline => UL_ALWAYS,
);

$mw->Show();
Win32::GUI::Dialog();

exit(0);

# using MessageBox for a quick About ... implementation
sub AboutWindow
{
	$mw->MessageBox("Demonstration of Win32::GUI::HyperLink v".Win32::GUI::HyperLink->VERSION()."\r\n".
		"Based on Win32::GUI v". Win32::GUI->VERSION(). "\r\n".
		'By Robert May - rmay@popeslane.clara.co.uk.',
		"About $title",
		MB_OK|MB_ICONASTERISK,
	);

	return 1;
}

sub setStatus
{
  my $self = shift;

  # An example of how the onMouseIn handler can be used to
  # change the window when the link is hovered over:
  # in this case, display the link in the status bar.
  $status->Text($self->Url());

  return 1;
}

sub unsetStatus
{
  my $self = shift;

  # An example of how the onMouseOut handler can be used to
  # change the window when the mouse leaves the link area:
  # in this case clearing the status bar text
  $status->Text("");

  return 1;
}

sub hl3Click
{
  my $self = shift;

  # An example of customising the behaviour when the link is clicked:
  # in this case, print a message, and then invoke the default handler.
  # You need to explicitly invoke the default handler if you want the
  # link to launch in your own handler.

  # print our message:
  print "You have clicked the link.\n";

  # invoke the default handler to launch the link:
  $self->Launch();

  return 1;
}
